[euvr] Subscription/Billing updates

vincentsalucci opened the pull request:
## Type of change

- [ ] Bug fix
- [X] New feature development
- [ ] Tech debt (refactoring, code cleanup, dependency upgrades, etc)
- [ ] Build/deploy pipeline (DevOps)
- [ ] Other

## Objective
> Update the `Settings` section to house a new `Subscription` menu option with the potential for `Go Premium/Premium Membership/ Payment Method / Billing History` tabs
> SG-19

## Code changes
- **organization-billing.component.html**: Copied over existing `user-billing.component.html` to maintain backwards compatibility for organizational settings (to be updated in the near future)
- **organization-billing.component.ts**: Removed base class and gave this component all the capabilities necessary to maintain backwards compatibility for organizational settings (to be updated in the near future)
- **oss-routing.module.ts**: Updated routing to match new design pattern
- **oss.module.ts**: Added newly created components
- **payment-method.component.html**: Created. Took useful pieces from the `user-billing.component`
- **payment-method.component.ts**: Created. Took useful pieces from the `user-billing.component`
- **premium.component.html**: Updated to match new design. Updated with new `CL` components where applicable
- **premium.component.ts**: Updated route to redirect when a user already has premium
- **settings.component.html**: Removed old navigation items and replaced with single `Subscription` option
- **settings.component.ts**: Added additional logic that will handle hiding the `Subscription` option under certain conditions.
- **subscription.component.html**: Created container for routing to selected tabs within the `Subscription` navigation option
- **subscription.component.ts**: Created. Handles supply of any logic needed for adjusting routes
- **subscription.module.ts**: Created. Handles lazy loading of child routes for the `Subscription` navigation option
- **user-billing.component.html**: Removed all `payment-method` related UI and updated with new `CL` components where applicable
- **user-billing..component.ts**: Removed all `payment-method` variables/logic/methods
- **user-subscription.component.html**: Adjusted for updated designs.  uUpdated with new `CL` components where applicable
- **user-subscription.component.ts**: Added helper method for title change. Updated navigation scheme on redirect.
- **messages.json**: Added `billingHistory`

## Screenshots
**Hidden Subscription due to proper conditions**
<img width="1032" alt="4-security-keys" src="https://user-images.githubusercontent.com/26154748/167161397-6b6dae84-98b7-481d-bc4d-d1cb1888891a.png">
<img width="1032" alt="5-sub-sub" src="https://user-images.githubusercontent.com/26154748/167161399-4937c3a5-0f40-4c60-86c6-30e4e7749cc1.png">
<img width="1032" alt="6-sub-payment" src="https://user-images.githubusercontent.com/26154748/167161402-21f21db7-d5e1-458f-8f18-4f1382668414.png">
<img width="1032" alt="7-sub-billing" src="https://user-images.githubusercontent.com/26154748/167161405-a064e68d-dafc-46c0-8517-c86e8541d247.png">
<img width="1032" alt="8-sub-self-host" src="https://user-images.githubusercontent.com/26154748/167161408-3a3dc347-39e6-456b-9642-1cbafdb9c3f8.png">

## Testing requirements
- Subscription flows for both cloud and self-hosted
- Organizational billing (should work as it currently does - need regression)

## Dependency
- [jslib - Separate Billing API](https://github.com/bitwarden/jslib/pull/750)

## Before you submit
- [X] I have checked for **linting** errors (`npm run lint`) (required)
- [ ] This change requires a **documentation update** (notify the documentation team)
- [ ] This change has particular **deployment requirements** (notify the DevOps team)

eliykat commented on src/app/settings/user-subscription.component.ts:
```diff
@@ -210,4 +210,8 @@ export class UserSubscriptionComponent implements OnInit {
   }
+
+  get title(): string {
+    return this.i18nService.t(this.selfHosted ? "subscription" : "premiumMembership");
```
Is this in the wireframes? I don't understand why we'd change the title this way.

vincentsalucci replied:
Good catch: It wasn't originally but the wireframes do have a header now. For the `Self Hosted` flow it shows `Subscription` instead of the current `Premium Membership` - I will double check with design today, since I technically did the headers before they were added into the wireframes. 👍

vincentsalucci replied:
I talked with design and the wireframes have been updated to reflect what is currently coded. This should be good to review 👍

eliykat commented on src/app/settings/user-subscription.component.ts:
```diff
@@ -210,4 +210,8 @@ export class UserSubscriptionComponent implements OnInit {
   }
+
+  get title(): string {
+    return this.i18nService.t(this.selfHosted ? "subscription" : "premiumMembership");
```
Is this in the wireframes? I don't understand why we'd change the title this way.

eliykat reviewed (changes requested):
Nice work! Just some smaller comments.

addisonbeck commented:
It looks like this PR is targeting `Feature/EndUserVaultRefresh`. There were some concerns raised from others about pascal case breaking some git clients. Can we change this to target `feature/endUserVaultRefresh`?

vincentsalucci commented:
> It looks like this PR is targeting `Feature/EndUserVaultRefresh`. There were some concerns raised from others about pascal case breaking some git clients. Can we change this to target `feature/endUserVaultRefresh`?

@addisonbeck - I saw that was mentioned in chat but had no idea we officially switched to `feature/endUserVaultRefresh`. I've also merged code into the now obsolete branch. I will cherry-pick to the new branch and try to re-point this one. Maybe that's why my build is broken /shrug

eliykat commented:
@vincentsalucci I'm making changes to these files in `master` for EC-74 - here's the diff so far: https://github.com/bitwarden/web/compare/feature/msp-hide-billing

This is going to cause a merge conflict as soon as we merge `feature/endUserVaultRefresh` into `master`, which will not have an obvious resolution.

What do you think about creating the standalone `organization-billing.component` in `master` first, as a separate PR? Then you can merge that change through into these feature branches, and I can safely make EC-74 changes to the org billing component without affecting EUVR work.

eliykat commented on src/app/organizations/settings/organization-billing.component.ts:
```diff
@@ -5,23 +5,36 @@ import { ApiService } from "jslib-common/abstractions/api.service";
 })
-export class OrganizationBillingComponent extends UserBillingComponent implements OnInit {
+export class OrganizationBillingComponent implements OnInit {
+  loading = false;
```
Can you add a `// TODO` somewhere in this file, noting that it should be combined with `payment-method.component` when we do the org admin refresh? I think it's reasonable to just copy + paste the logic for now so that we don't go out of scope, but I want to make sure it's reconsidered in the next cycle.

vincentsalucci replied:
I'm not sure I follow - the org admin refresh will follow the same pattern as the current refresh and separate billing/invoice and payment method/balance.

eliykat replied:
Basically, there's a lot of copy + pasted logic duplicated between `organization-billing.component.ts` and `user-billing.component.ts` as a result of these changes. My understanding is that this is temporary to preserve the existing behaviour of the org pages. My thinking is, we want some flag here to make sure this duplicated code doesn't hang around. Otherwise it should be moved to a base class so it can be shared between components.

vincentsalucci replied:
It is only temporary and there will be no duplicated code once the org admin refresh takes place as the `organization-billing.component.ts` will be refactored to extend the `user-billing-history.component.ts` as it exists in production today. Will add a `TODO` as suggested as another reminder!

eliykat replied:
Yes I agree - we're on the same page, just wanted a TODO in there as a reminder, once we've moved on and dumped all of this out of our heads :D

vincentsalucci commented:
> @vincentsalucci I'm making changes to these files in `master` for EC-74 - here's the diff so far: https://github.com/bitwarden/web/compare/feature/msp-hide-billing
> 
> This is going to cause a merge conflict as soon as we merge `feature/endUserVaultRefresh` into `master`, which will not have an obvious resolution.
> 
> What do you think about creating the standalone `organization-billing.component` in `master` first, as a separate PR? Then you can merge that change through into these feature branches, and I can safely make EC-74 changes to the org billing component without affecting EUVR work.

@eliykat I will work on this asap to try and get it to you soon ™️

vincentsalucci commented on src/app/settings/settings.component.ts:
```diff
@@ -47,5 +53,8 @@ export class SettingsComponent implements OnInit, OnDestroy {
   async load() {
     this.premium = await this.tokenService.getPremium();
     this.hasFamilySponsorshipAvailable = await this.organizationService.canManageSponsorships();
+    const hasPremiumFromOrg = await this.stateService.getCanAccessPremium();
```
Ah, I see - I did the stretch goal task code after I took the original screenshots. I apologize and see how that's confusing :D  Will add a screenshot of the `Subscription` vertical navigation tab being hidden as expected 👍

vincentsalucci replied:
Updated the screenshots to the latest changes for you @eliykat

vincentsalucci commented on src/app/settings/settings.component.ts:
```diff
@@ -1,10 +1,13 @@
 import { PlatformUtilsService } from "jslib-common/abstractions/platformUtils.service";
 import { TokenService } from "jslib-common/abstractions/token.service";
 
+import { StateService } from "src/abstractions/state.service";
```
It is! Just a bit confusing because this is actually a web specific `StateService` that allows for use of additional abstracted methods. It extends our `jslib` `StateService` abstraction within.

vincentsalucci commented on src/app/settings/premium.component.html:
```diff
@@ -1,15 +1,21 @@
-<div class="page-header">
-  <h1>{{ "goPremium" | i18n }}</h1>
+<div
```
Updated!

vincentsalucci commented on src/app/settings/user-billing.component.html:
```diff
@@ -1,10 +1,12 @@
-<div class="page-header d-flex">
```
Updated to `user-billing-history`!

eliykat commented:
> > @vincentsalucci I'm making changes to these files in `master` for EC-74 - here's the diff so far: https://github.com/bitwarden/web/compare/feature/msp-hide-billing
> > This is going to cause a merge conflict as soon as we merge `feature/endUserVaultRefresh` into `master`, which will not have an obvious resolution.
> > What do you think about creating the standalone `organization-billing.component` in `master` first, as a separate PR? Then you can merge that change through into these feature branches, and I can safely make EC-74 changes to the org billing component without affecting EUVR work.
> 
> @eliykat I will work on this asap to try and get it to you soon ™️

Sorry, turns out I won't actually be touching this component, so you don't need to worry about this extra work. Please proceed with the original PR as you were :)

vincentsalucci commented on src/app/settings/user-billing.component.html:
```diff
@@ -1,10 +1,12 @@
+    buttonType="secondary"
     (click)="load()"
-    class="btn btn-sm btn-outline-primary ml-auto"
+    class="ml-auto"
```
Updated!

eliykat reviewed (commented):
2 differences I noticed from the wireframes - not sure if it's intentional:

1. The wireframe has the Premium Membership tab laid out more vertically:

![Screen Shot 2022-05-05 at 3 08 43 pm](https://user-images.githubusercontent.com/31796059/166866535-2a6348ec-f655-4ce8-a9fe-6459fc2f5cc1.png)

Compare to the current layout:

![Screen Shot 2022-05-05 at 3 08 36 pm](https://user-images.githubusercontent.com/31796059/166866549-b9900dc8-6647-4677-9a3c-7e49d5e732c9.png)

2. You have a divider line under your headings, whereas the Figma doesn't - also visible in the screenshots above. The reason I noticed is that it starts to carve up the page a little especially in Billing History:

![Screen Shot 2022-05-05 at 3 12 32 pm](https://user-images.githubusercontent.com/31796059/166866672-2e6b64e3-28e0-4503-b8a9-a89bce570d78.png)

If design says it's fine then it's fine, just checking.

eliykat commented on src/app/settings/user-subscription.component.html:
```diff
@@ -42,7 +42,6 @@ <h1>
       bit-button
       buttonType="secondary"
       #reinstateBtn
-      type="button"
```
Any reason for removing this? We should always specify `type` for buttons, and `type="button"` is a good default. Otherwise it can default to a submit button and cause unexpected behaviour.

eliykat replied:
Same comment for all other buttons on this page where you've removed the `type` attribute.

Hinton replied:
We should omit `type="button"` on `button` elements unless it's in a form, in which case we need type to not make it a submit button.

eliykat replied:
That seems reasonable, it is form behaviour that I was concerned about. I couldn't find the PR you refer to, what's the accessibility reason? Is it read out by screenreaders or something?

Hinton replied:
Seems to be my bad. I had confused `role` and `type`. Let's keep `type="button"` to be safe.

vincentsalucci replied:
Added `type="button"`back to the button declarations

vincentsalucci commented:
> 2 differences I noticed from the wireframes - not sure if it's intentional:
> 
> 1. The wireframe has the Premium Membership tab laid out more vertically:
> 
> ![Screen Shot 2022-05-05 at 3 08 43 pm](https://user-images.githubusercontent.com/31796059/166866535-2a6348ec-f655-4ce8-a9fe-6459fc2f5cc1.png)
> 
> Compare to the current layout:
> 
> ![Screen Shot 2022-05-05 at 3 08 36 pm](https://user-images.githubusercontent.com/31796059/166866549-b9900dc8-6647-4677-9a3c-7e49d5e732c9.png)
> 
> 2. You have a divider line under your headings, whereas the Figma doesn't - also visible in the screenshots above. The reason I noticed is that it starts to carve up the page a little especially in Billing History:
> 
> ![Screen Shot 2022-05-05 at 3 12 32 pm](https://user-images.githubusercontent.com/31796059/166866672-2e6b64e3-28e0-4503-b8a9-a89bce570d78.png)
> 
> If design says it's fine then it's fine, just checking.

1. This is my fault overall - but since this wasn't mentioned at all in the task breakdown, I completely overlooked the new layout, but I'm also not sure its intentional as this is pretty far from what's in place today. I will check-in with design.

2. Fixed this - I made the changes while the builds were broken and wasn't able to test it until now. I've updated the related `sass` to remove dividers for tabbed pages and also adjusted the margins to closer resemble the wireframes.

vincentsalucci commented:
@eliykat Talked with design - they've updated the wireframes to match what we currently have in place.

Hinton commented on src/app/settings/user-billing-history.component.ts:
```diff
@@ -0,0 +1,53 @@
+})
+export class UserBillingHistoryComponent implements OnInit {
+  loading = false;
+  firstLoaded = false;
```
We should really look into adopting https://angular.io/api/router/Resolve, these loading page behavior is bugging me.

Hinton commented on src/app/settings/user-billing-history.component.ts:
```diff
@@ -0,0 +1,53 @@
+  ) {}
+
+  async ngOnInit() {
+    if (this.platformUtilsService.isSelfHost()) {
+      this.router.navigate(["/settings/subscription"]);
+    }
```
Feels like this belongs better in a guard class of some sort. Creating a tech debt item to figure out this kind of redirects.

Hinton commented on src/app/organizations/settings/organization-billing.component.ts:
```diff
@@ -30,4 +45,110 @@ export class OrganizationBillingComponent extends UserBillingComponent implement
+      return;
+    }
+    this.loading = true;
+    if (this.organizationId != null) {
+      this.billing = await this.apiService.getOrganizationBilling(this.organizationId);
+    }
```
Can we enter this component without an orgnaizationId?

vincentsalucci replied:
After reviewing the routing module/permissions for this component, it appears the answer to your question is no. This is just leftover for temporarily porting over the super class until the admin vault refresh.

Hinton replied:
@vincentsalucci can you ensure we have a ticket for to ensure this gets completed in the admin refresh?

Hinton commented on src/app/settings/payment-method.component.html:
```diff
@@ -0,0 +1,69 @@
+  <p *ngIf="!paymentSource">{{ "noPaymentMethod" | i18n }}</p>
+  <ng-container *ngIf="paymentSource">
+    <p>
+      <i
+        class="bwi bwi-fw"
+        [ngClass]="{
+          'bwi-credit-card': paymentSource.type === paymentMethodType.Card,
+          'bwi-bank': paymentSource.type === paymentMethodType.BankAccount,
+          'bwi-money': paymentSource.type === paymentMethodType.Check,
+          'bwi-paypal text-primary': paymentSource.type === paymentMethodType.PayPal,
+          'bwi-apple text-muted': paymentSource.type === paymentMethodType.AppleInApp,
+          'bwi-google text-muted': paymentSource.type === paymentMethodType.GoogleInApp
+        }"
+      ></i>
```
We're duplicating this in two places, we should add a tech debt item to create a component for this.

vincentsalucci replied:
This code duplication is only temporary and will be addressed with the Admin Vault Refresh.

Hinton commented on src/app/settings/premium.component.html:
```diff
@@ -1,15 +1,18 @@
-<div class="page-header">
+<div *ngIf="selfHosted" class="page-header">
+  <h1>{{ "subscription" | i18n }}</h1>
+</div>
+<div *ngIf="!selfHosted" class="tabbed-header">
```
We are using a different pattern here than in `src/app/settings/user-subscription.component.html`. We should pick one.

vincentsalucci replied:
That's because of the pre-existing loading pattern that's in place for the `user-subscription` component. I didn't feel that it was within scope to try and restructure that for the sole purpose of supplying a header text.

Hinton replied:
@vincentsalucci can you ensure we have a tech debt ticket for this in the SG pod?

Hinton reviewed (approved):
Approving this, assuming that we tackle the comments in a tech debt item afterwards.
